Guide.docc/IncrementalAdoption.md

```diff
@@ -1,3 +1,236 @@
+[continuation-apis]: https://developer.apple.com/documentation/swift/concurrency#continuations
+
+## Dynamic Isolation
+
+Expressing the isolation of your program statically, using annotations and
+other language constructs, is both powerful and concise.
+But, because static isolation uses Swift's type system, introducing it
+can potentially require changes to all places a type is used.
+
+Dynamic isolation provides runtime mechanisms for describing data isolation.
+These are a more manual, but also more flexible alternative.
+Dynamic isolation can be an essential tool for interfacing a Swift 6 component
+with another that has not yet migrated,
+even if these components are within the _same_ module.
+
+### Internal-Only Isolation
+
+Suppose you have determined that a reference type within your project can be
+best described with `MainActor` static isolation.
+
+```swift
+@MainActor
+class WindowStyler {
+    private var backgroundColor: ColorComponents
+
+    func applyStyle() {
+        // ...
+    }
+}
+```
+
+This `MainActor` isolation may be _logically_ correct.
+But, if this type is used in other unmigrated locations,
+adding static isolation here could require many additional changes.
+An alternative is to use dynamic isolation to help control the scope.
+
+```swift
+class WindowStyler {
+    @MainActor
+    private var backgroundColor: ColorComponents
+
+    func applyStyle() {
+        MainActor.assumeIsolated {
+            // use and interact with other `MainActor` state
+        }
+    }
+}
+```
+
+Here, the isolation has been internalized into the class.
+This keeps any changes localized to the type, allowing you make
+changes without affecting any clients of the type.
+
+However, a disadvantage of this technique is now the type and its internals
+do not have the same isolation domain.
+This can result in non-`Sendable` method parameters or return values crossing
+otherwise-unnecessary isolation boundaries.
+
+### Usage-Only Isolation
+
+If it is impractical to contain isolation exclusively within a type, you can
+instead expand the isolation to cover only its API usage.
+
+To do this, first apply static isolation to the type,
+and then use dynamic isolation at any usage locations:
+
+```swift
+@MainActor
+class WindowStyler {
+    // ...
+}
+
+class UIStyler {
+    @MainActor
+    private let windowStyler: WindowStyler
+    
+    func applyStyle() {
+        MainActor.assumeIsolated {
+            windowStyler.applyStyle()
+        }
+    }
+}
+```
+
+A variant of this technique can also work in an asynchronous context.
+This is particularly helpful when you have a class that already has
+`async` methods, but hasn't yet established correct static isolation.
+
+```swift
+@MainActor
+class WindowStyler {
+    // ...
+}
+
+class UIStyler {
+    @MainActor
+    private let windowStyler = WindowStyler()
+
+    func updateStyle() async {
+        await MainActor.run {
+            windowStyler.applyStyle()
```

Comment:
I'm a bit confused. Why use `MainActor.run` instead of just `await windowStyler.applyStyle()`? (I don't like `MainActor.run` for this very reason - it seems like you need it when you don't).

Reply:
🙈